core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php

```diff
@@ -0,0 +1,244 @@
+        $currentToken = $this->tokenStorage->getToken();
+
+        if ($currentToken instanceof TwoFactorTokenInterface) {
+            return $this->twoFactorAuthenticator->authenticate($request);
+        }
+
+        $credentials = $this->getCredentials($request);
+
+        $passport = new Passport(
+            new UserBadge($credentials['username'], [$this->userProvider, 'loadUserByIdentifier']),
+            new PasswordCredentials($credentials['password']),
+            [new RememberMeBadge()]
+        );
+
+        if ($this->options['enable_csrf']) {
+            $passport->addBadge(new CsrfTokenBadge($this->options['csrf_token_id'], $credentials['csrf_token']));
+        }
+
+        if ($this->userProvider instanceof PasswordUpgraderInterface) {
+            $passport->addBadge(new PasswordUpgradeBadge($credentials['password'], $this->userProvider));
+        }
+
+        return $passport;
+    }
+
+    public function createToken(Passport $passport, string $firewallName): TokenInterface
+    {
+        $credentialsBadge = $passport->getBadge(TwoFactorCodeCredentials::class);
+
+        if (!$credentialsBadge instanceof TwoFactorCodeCredentials) {
+            return parent::createToken($passport, $firewallName);
+        }
+
+        $twoFactorToken = $credentialsBadge->getTwoFactorToken();
+
+        if ($twoFactorToken->allTwoFactorProvidersAuthenticated()) {
+            $authenticatedToken = $twoFactorToken->getAuthenticatedToken(); // Authentication complete, unwrap the token
+            $authenticatedToken->setAttribute(TwoFactorAuthenticator::FLAG_2FA_COMPLETE, true);
+
+            return $authenticatedToken;
+        }
+
+        return $twoFactorToken;
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
+    {
+        return $this->successHandler->onAuthenticationSuccess($request, $token);
+    }
+
+    public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
+    {
+        return $this->failureHandler->onAuthenticationFailure($request, $exception);
+    }
+
+    public function isInteractive(): bool
+    {
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (null === $request) {
+            return false;
+        }
+
+        $page = $request->attributes->get('pageModel');
+
+        if (!$page instanceof PageModel) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private function getCredentials(Request $request): array
+    {
+        $credentials = [
+            'csrf_token' => ParameterBagUtils::getRequestParameterValue($request, $this->options['csrf_parameter']),
+            'username' => ParameterBagUtils::getParameterBagValue($request->request, $this->options['username_parameter']),
+            'password' => ParameterBagUtils::getParameterBagValue($request->request, $this->options['password_parameter']) ?? '',
```

Comment:
<img width="893" alt="" src="https://user-images.githubusercontent.com/1192057/178517313-cfec0f72-d172-4dbf-9a42-71ca70224d7e.png">

I assume there is no other way, is there?

Reply:
As discussed with @leofeyer we will keep it as it is and tackle/discuss the `@internal` usages all over Contao in a different PR.